source/log.c

```diff
@@ -0,0 +1,302 @@
+    }
+
+    aws_atomic_add(&s_log_thread_count, 1);
+    aws_atomic_set_ptr((void **)&s_local_log_context, aws_mem_acquire(alloc, sizeof(struct aws_log_context)));
+    if (!s_local_log_context) {
+        aws_raise_error(AWS_ERROR_OOM);
+        return AWS_OP_ERR;
+    }
+
+    aws_secure_zero(s_local_log_context, sizeof(*s_local_log_context));
+    s_local_log_context->max_message_len = max_message_len;
+    s_local_log_context->alloc = alloc;
+    s_local_log_context->message_pool = aws_memory_pool_init(alloc, sizeof(struct aws_log_message) + max_message_len, memory_pool_message_count);
+    if (!s_local_log_context->message_pool) {
+        aws_raise_error(AWS_ERROR_OOM);
+        return AWS_OP_ERR;
+    }
+    AWS_CIRCULAR_DOUBLY_LIST_INIT(s_local_log_context);
+
+    /* Insert thread local log info struct onto the global doubly linked list used by
+    the logging thread. */
+    aws_mutex_lock(&s_log_list_mutex);
+    if (!s_log_list) {
+        s_log_list = s_local_log_context;
+    } else {
+        AWS_CIRCULAR_DOUBLY_LIST_INSERT_BEFORE(s_log_list, s_local_log_context);
+    }
+    aws_mutex_unlock(&s_log_list_mutex);
+
+
+    return AWS_OP_SUCCESS;
+}
+
+int aws_log_clean_up() {
+    if (aws_atomic_get(&s_log_thread_count) == 0) {
+        aws_raise_error(AWS_ERROR_LOG_UNINITIALIZED);
+        return AWS_OP_ERR;
+    }
+
+    struct aws_log_context *context;
+    do {
+        context = s_local_log_context;
+    } while (!aws_atomic_cas_ptr((void **)&s_local_log_context, context, NULL));
+
+    /* Perform final cleanup if this is the last log thread. */
+    aws_mutex_lock(&s_log_list_mutex);
+    if (aws_atomic_add(&s_log_thread_count, -1) == 1) {
+        struct aws_log_context *log_list = s_log_list;
+        s_log_list = NULL;
+        aws_mutex_unlock(&s_log_list_mutex);
+
+        /* Make sure all the other log threads have been joined and remove dead threads. */
+        struct aws_log_context *sentinel = log_list;
+        do {
+            struct aws_log_context *next = log_list->next;
+
+            /* Remove dead threads. Log any leftover messages. Cleanup leftover message pool memory. */
+            struct aws_log_message *delete_list = log_list->message_list;
+            while (delete_list) {
+                struct aws_log_message *next0 = delete_list->next;
+                char *msg_data = (char*)(delete_list + 1);
+                s_log_report_callback(msg_data);
+                aws_memory_pool_release(log_list->message_pool, delete_list);
+                delete_list = next0;
+            }
+
+            delete_list = log_list->delete_list;
+            while (delete_list) {
+                struct aws_log_message *next0 = delete_list->next;
+                aws_memory_pool_release(log_list->message_pool, delete_list);
+                delete_list = next0;
+            }
+
+            aws_memory_pool_clean_up(log_list->message_pool);
+            aws_mem_release(log_list->alloc, log_list);
+

```

Comment:
I suggest adding `aws_secure_zero(log_list);` here to avoid any sensitive information that was logged from lingering in memory.